mino/minogrpc/server_test.go

```diff
@@ -785,6 +790,327 @@ func TestRPC_MultipleRing_Stream(t *testing.T) {
+	}
+	server3, err := CreateServer(addr3)
+	require.NoError(t, err)
+	server3.traffic.log = true
+	server3.StartServer()
+	peer3 := Peer{
+		Address:     server3.listener.Addr().String(),
+		Certificate: server3.cert.Leaf,
+	}
+
+	identifier4 := "127.0.0.1:2004"
+	addr4 := &address{
+		id: identifier4,
+	}
+	server4, err := CreateServer(addr4)
+	require.NoError(t, err)
+	server4.traffic.log = true
+	server4.StartServer()
+	peer4 := Peer{
+		Address:     server4.listener.Addr().String(),
+		Certificate: server4.cert.Leaf,
+	}
+
+	// We have the following topology:
+	// srv1:
+	// - srv2 (direct)
+	// - srv3 (via srv2)
+	// - srv4 (via srv2)
+	// srv2:
+	// - srv3 (direct)
+	// - srv4 (via srv3)
+	// srv3:
+	// - srv4 (direct)
+	server1.neighbours[identifier1] = peer1
+	server1.neighbours[identifier2] = peer2
+	server1.routingTable[identifier3] = identifier2
+	server1.routingTable[identifier4] = identifier2
+
+	server2.mesh[identifier3] = peer3
+	server2.routingTable[identifier4] = identifier3
+
+	server3.mesh[identifier4] = peer4
+
+	uri := "blabla"
+	handler1 := testMeshHandler{addrID: identifier1}
+	rpc := RPC{
+		handler: handler1,
+		srv:     server1,
+		uri:     uri,
+	}
+
+	// the handler must be registered on each server. Fron the client side, that
+	// means the "registerNamespace" and "makeRPC" must be called on each
+	// server.
+	server1.handlers[uri] = handler1
+	server2.handlers[uri] = testMeshHandler{addrID: identifier2}
+	server3.handlers[uri] = testMeshHandler{addrID: identifier3}
+	server4.handlers[uri] = testMeshHandler{addrID: identifier4}
+
+	dummyMsg := &wrappers.StringValue{Value: "dummy_value"}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	sender, rcvr := rpc.Stream(ctx, &fakePlayers{players: []address{*addr1, *addr2, *addr3, *addr4}})
+	localRcvr, ok := rcvr.(receiver)
+	require.True(t, ok)
+
+	select {
+	case err := <-localRcvr.errs:
+		t.Errorf("unexpected error in rcvr: %v", err)
+	case <-time.After(time.Millisecond * 200):
+	}
+
+	// sending message to server1, which should send to its neighbor, etc...
+	errs := sender.Send(dummyMsg, addr1, addr2, addr3, addr4)
+	err, more := <-errs
+	if more {
+		t.Error("unexpected error from send: ", err)
+	}
+
+	msgs := make([]string, 4)
+
+	for i := range msgs {
+		_, msg2, err := rcvr.Recv(context.Background())
+		require.NoError(t, err)
+
+		dummyMsg2, ok := msg2.(*wrappers.StringValue)
+		require.True(t, ok)
+		msgs[i] = dummyMsg2.Value
+	}
+	sort.Strings(msgs)
+	require.Equal(t, "dummy_value_"+identifier1, msgs[0])
+	require.Equal(t, "dummy_value_"+identifier2, msgs[1])
+	require.Equal(t, "dummy_value_"+identifier3, msgs[2])
+	require.Equal(t, "dummy_value_"+identifier4, msgs[3])
+
+	// fmt.Println("server1: " + server1.traffic.String())
+	// fmt.Println("server2: " + server2.traffic.String())
+	// fmt.Println("server3: " + server3.traffic.String())
+	// fmt.Println("server4: " + server4.traffic.String())
+
+	// // traffic of server 1
```

Comment:
You forgot to uncomment ?

Reply:
got some trouble with having deterministic behaviour, making checking the logs impossible.
I'm wondering if we should have a flag that disable asynchronous calls: That would allow better testing and debugging.